File: Sources/SomnioEditor/ViewModels/CursorReadout.swift
import Foundation
import SomnioCore

/// HUD strip values: X/Y track the cursor in sector-pixel coordinates (top-left origin)
/// and W/H mirror the bounds of the currently-selected record. Three of the four values
/// update from continuous-hover events; W/H comes from `.onChange(of: workspace.selection)`
/// because the `@Observable` macro discards user-provided `didSet`/`willSet` accessors.
@Observable public final class CursorReadout {
    public var x: Int16 = 0
    public var y: Int16 = 0
    public var width: Int16 = 0
    public var height: Int16 = 0

    public init() {}

    /// Pulls the selected record's bounds out of the sector body. Called whenever the
    /// selection changes; W/H track a single selection only (a multi-selection has no one
    /// record size) and clear when nothing is selected so the readout doesn't stay pinned
    /// to a deleted record's old dimensions.
    public func applyBounds(for selection: Set<EditorSelection>, in body: SectorBody) {
        guard selection.count == 1, let size = selection.first?.bounds(in: body)?.size else {
            width = 0
            height = 0
            return
        }
        width = size.width
        height = size.height
    }
}

File: Web/src/editor/ui/cursorReadout.ts
import { element } from '@/ui/dom'
import type { Sector } from '@/core/sector'
import { selectionBounds } from '../selection'
import type { EditorSelection } from '../selection'

/**
 * The bottom-leading status strip: X/Y track the hovered sector pixel, W/H mirror the single
 * selection's bounds (a multi-selection has no one record size, so they clear), plus the
 * sector name.
 */
export class CursorReadout {
  readonly root: HTMLElement

  x = 0
  y = 0
  width = 0
  height = 0

  private readonly text = element('span')

  constructor() {
    this.root = element('div', {
      className: 'fantasy-panel editor-readout',
      children: [this.text],
    })
    this.render('')
  }

  /** W/H track a single selection only, and clear when nothing (or several) is selected. */
  applyBounds(selection: readonly EditorSelection[], sector: Sector): void {
    if (selection.length !== 1) {
      this.width = 0
      this.height = 0
      return
    }
    const bounds = selectionBounds(selection[0]!, sector)
    this.width = bounds?.size.width ?? 0
    this.height = bounds?.size.height ?? 0
  }

  render(sectorName: string): void {
    const name = sectorName === '' ? '' : `  ${sectorName}`
    this.text.textContent = `X: ${this.x}  Y: ${this.y}  W: ${this.width}  H: ${this.height}${name}`
  }
}
